text/0000-force-declaration-of-ffi-functions.md

```diff
@@ -0,0 +1,214 @@
+  26 ./net_ssl/net_ssl/ssl.pony
+  26 ./net_ssl/net_ssl/ssl_context.pony
+  18 ./net_ssl/net_ssl/x509.pony
+   1 ./reactive_streams/examples/spl4/main.pony
+   9 ./regex/regex/match.pony
+  11 ./regex/regex/regex.pony
+```
+
+# How We Teach This
+
+- The ["C FFI" chapter of the tutorial](https://tutorial.ponylang.io/c-ffi.html) should be updated to reflect this change.
+
+    In particular, the ["Calling C from Pony" section](https://tutorial.ponylang.io/c-ffi/calling-c.html) should note that a declaration is needed in order to call a FFI function, as well as introduce the `...` syntax for declaring variadic functions. Although we could expect that users that attempt to use the C FFI will be familiar with the concept of variadic functions, it wouldn't hurt to mention them, even if we don't use the "variadic" term.
+
+    The chapter should also remove any explicit type declarations from FFI calls.
+
+- All the examples that show FFI code, be it in the `examples/` folder of the compiler or on the tutorial should be changed to include mandatory declarations, and remove return types at the call site. These examples are located in the C FFI and [Serialisation](https://tutorial.ponylang.io/appendices/serialisation.html) chapters, and in the [examples appendix](https://tutorial.ponylang.io/appendices/examples.html).
+
+In general, the use of the C-FFI should be considered an advanced feature, as its usage can violate many of the safety properties enforced by the language. As such, not much would change for regular Pony users.
+
+# How We Test This
+
+* All currently supported platforms have common calling conventions for regular and variadic functions. The change in code generation shouldn't affect any of these platforms.
+
+* The use of the appropriate calling conventions will be handled by LLVM. We would assume that LLVM is able to do this correctly for all supported platforms.
+
+* Compiler tests can/should be written to demonstrate that programs that lack explicit FFI declarations fail to compile.
+
+* Apple ARM64 should be added as a target platform for CI when our current provider (Cirrus CI) offers instances of this type. In the meantime, the Pony team owns two machines with Apple Silicon chips that should serve as testing infrastructure.
+
+* We should ensure that `use` scoping rules allow us to remove explicit return types on FFI calls.
+
+# Drawbacks
+
+* Breaks existing code that uses FFI functions.
+
+* Imposes overhead on users to declare the type of FFI functions in advance.
+
+# Alternatives
+
+* Make the compiler treat FFI functions as regular functions by default, and require explicit declaration only for variadic FFI functions.
+
+* Use an alternative syntax to let the compiler know about variadic functions at the call site, so that a declaration is not needed, such as (for example) `@printf[I32](...)(arg1, arg2)`.
+
+* Leave the compiler as it is, and do not support platforms with different calling conventions.
+
```

Comment:
One other alternative is to be able to handle header files from C in some way for this use case. Would keep it convenient for most use cases

Reply:
I think that's a good add on feature for FFI. I'm not sure I would call it an alternative given what a large change it would be from the existing FFI mechanisms.

Reply:
My thought is that it could be used to give enough information to continue handling FFI without "use declarations"

Reply:
What's really nice about the current is that I can compile for any supported platform and it will work and type check without needing any additional headers etc.

I've never installed windows headers on my machine but the declarations give me type checking.

I love that if a pony program type checks on one platform, it type checks on all.

I don't want to lose that.

The FFI declarations are "trust me" for that type checking but I still LOVE what it gives me and do not want to give it up.

Reply:
The selfish part of me would love to be able to take C header files in lieu of FFI declaration proposals because I spend a non-trivial amount of time converting header files into these types of declarations.  It's non-trivial both because of the number of functions I'm dealing with, but also the number of macros and other cpp tomfoolery that's found in many C libraries.

With 'self-authored' FFI declarations it's type-checking my function call against the what I understand the C function prototype to be.

If it's generated from the header-file, then it removes human error from another sensitive part of the process which is a good thing for software reliability - especially when juggling knives with C-FFI (by definition) at runtime.

In order to not lose the platform independence functionality that you value so much @SeanTAllen - would/could/should an independent tool which converts C header files into use statements exist?

Then potentially we could have the best of both worlds.

Reply:
(How one would go about writing such a thing, (and if it's even feasible) is left as an exercise for the reader?)

Reply:
An independent tool could do that. I think it should be optional.

I shouldn't be required to have headers installed to type check.

I don't think there is anything preventing folks from developing such a tool now.

Reply:
(arguably, this is the same hard problem that I've been struggling with for the last month or so)

Reply:
Another thing we currently have is you can support multiple API versions of the same library like OpenSSL and type check them all in one go.

That's built on the same foundation as the "type check all platforms" functionality I mentioned earlier.

Reply:
The thing about the approach of scanning C headers / type signatures, in addition to the extra compiler complexity and the header dependency problem @SeanTAllen mentioned, is that it does not necessarily give enough information to wrap the calls for use in another language. There are various cases of ambiguity that need an extra information channel (such as human-interpreted documentation, or something like the GIntrospection format that @redvers has been using) to resolve to the most appropriate Pony FFI use patterns.

Reply:
I agree that this would be a useful feature for Pony, but I have to side with Sean, I don't think it should be a requirement to have headers installed to use the FFI. Right now I can write some windows-dependent code on unix* platforms by reading the reference on msdn, and the compiler will check the types. It's not a guarantee that the code will work, but I'd like to keep that option open.